kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorRateLimiter.java

```diff
@@ -107,25 +107,100 @@ protected static RateLimiter smoothBurstyRateLimiter(
     }
   }
 
+  public static int getSpanBytes(List<Trace.Span> spans) {
+    return spans.stream().mapToInt(Trace.Span::getSerializedSize).sum();
+  }
+
+  public Predicate<String, List<Trace.Span>> createBulkIngestRateLimiter(
```

Comment:
This is likely going to need some considerable refactoring once we switch to http-only. There's a few issues I can see with this I'm not a huge fan of:

* This still uses a Predicate, which is a Kafka Streams abstraction
* This current rate limiter doesn't have a transactional abstraction. This means that when attempting to reserve rate limits for a set of services, if any fail we don't return the unused rate limits when we fail the request. 

I don't think the second issue needs to be addressed in this PR, but we might want to see if the first is easy to resolve.

Reply:
addressed 1 with BiPredicate